File: src/main/java/fr/adrienbrault/idea/symfony2plugin/dic/inspection/XmlServiceInstanceInspection.java
package fr.adrienbrault.idea.symfony2plugin.dic.inspection;

import com.intellij.codeInspection.LocalInspectionTool;
import com.intellij.codeInspection.ProblemHighlightType;
import com.intellij.codeInspection.ProblemsHolder;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiElementVisitor;
import com.intellij.psi.util.PsiTreeUtil;
import com.intellij.psi.xml.XmlAttribute;
import com.intellij.psi.xml.XmlAttributeValue;
import com.intellij.psi.xml.XmlTag;
import com.jetbrains.php.lang.psi.elements.Method;
import com.jetbrains.php.lang.psi.elements.Parameter;
import com.jetbrains.php.lang.psi.elements.PhpClass;
import fr.adrienbrault.idea.symfony2plugin.Symfony2ProjectComponent;
import fr.adrienbrault.idea.symfony2plugin.config.xml.XmlHelper;
import fr.adrienbrault.idea.symfony2plugin.dic.container.suggestion.XmlServiceSuggestIntentionAction;
import fr.adrienbrault.idea.symfony2plugin.stubs.ContainerCollectionResolver;
import fr.adrienbrault.idea.symfony2plugin.util.PhpElementsUtil;
import fr.adrienbrault.idea.symfony2plugin.util.dict.ServiceUtil;
import org.apache.commons.lang.StringUtils;
import org.jetbrains.annotations.NotNull;

/**
 * @author Daniel Espendiller <daniel@espendiller.net>
 */
public class XmlServiceInstanceInspection extends LocalInspectionTool {
    @NotNull
    public PsiElementVisitor buildVisitor(final @NotNull ProblemsHolder holder, boolean isOnTheFly) {
        if (!Symfony2ProjectComponent.isEnabled(holder.getProject())) {
            return super.buildVisitor(holder, isOnTheFly);
        }

        return new PsiElementVisitor() {
            private ContainerCollectionResolver.LazyServiceCollector lazyServiceCollector;

            @Override
            public void visitElement(@NotNull PsiElement psiElement) {
                if(XmlHelper.getArgumentServiceIdPattern().accepts(psiElement)) {
                    if (this.lazyServiceCollector == null) {
                        this.lazyServiceCollector = new ContainerCollectionResolver.LazyServiceCollector(holder.getProject());
                    }

                    annotateServiceInstance(psiElement, holder, this.lazyServiceCollector);
                }

                super.visitElement(psiElement);
            }
        };
    }

    private void annotateServiceInstance(@NotNull PsiElement psiElement, @NotNull ProblemsHolder holder, @NotNull ContainerCollectionResolver.LazyServiceCollector lazyServiceCollector) {
        // search for parent service definition
        XmlTag currentXmlTag = PsiTreeUtil.getParentOfType(psiElement, XmlTag.class);
        XmlTag parentXmlTag = PsiTreeUtil.getParentOfType(currentXmlTag, XmlTag.class);
        if(parentXmlTag == null) {
            return;
        }

        String name = parentXmlTag.getName();
        if(name.equals("service")) {
            // service/argument[id]
            String serviceDefName = XmlHelper.getClassFromServiceDefinition(parentXmlTag);
            if(serviceDefName != null) {
                PhpClass phpClass = ServiceUtil.getResolvedClassDefinition(holder.getProject(), serviceDefName, lazyServiceCollector);

                // check type hint on constructor
                if(phpClass != null) {
                    Method constructor = phpClass.getConstructor();
                    if(constructor != null) {
                        String serviceName = ((XmlAttributeValue) psiElement).getValue();
                        if(StringUtils.isNotBlank(serviceName)) {
                            attachMethodInstances(psiElement, serviceName, constructor, XmlHelper.getArgumentIndex(currentXmlTag, constructor), holder, lazyServiceCollector);
                        }
                    }
                }
            }
        } else if (name.equals("call")) {

            // service/call/argument[id]

            XmlAttribute methodAttribute = parentXmlTag.getAttribute("method");
            if(methodAttribute != null) {
                String methodName = methodAttribute.getValue();
                XmlTag serviceTag = parentXmlTag.getParentTag();

                // get service class
                if(serviceTag != null && "service".equals(serviceTag.getName())) {
                    String serviceDefName = XmlHelper.getClassFromServiceDefinition(serviceTag);
                    if(serviceDefName != null) {
                        PhpClass phpClass = ServiceUtil.getResolvedClassDefinition(holder.getProject(), serviceDefName, lazyServiceCollector);

                        // finally check method type hint
                        if(phpClass != null) {
                            Method method = phpClass.findMethodByName(methodName);
                            if(method != null) {
                                String serviceName = ((XmlAttributeValue) psiElement).getValue();
                                if(StringUtils.isNotBlank(serviceName)) {
                                    attachMethodInstances(psiElement, serviceName, method, XmlHelper.getArgumentIndex(currentXmlTag, method), holder, lazyServiceCollector);
                                }
                            }
                        }
                    }
                }
            }
        }
    }

    private void attachMethodInstances(@NotNull PsiElement target, @NotNull String serviceName, @NotNull Method method, int parameterIndex, @NotNull ProblemsHolder holder, ContainerCollectionResolver.@NotNull LazyServiceCollector lazyServiceCollector) {
        Parameter[] constructorParameter = method.getParameters();
        if(parameterIndex >= constructorParameter.length) {
            return;
        }

        String className = constructorParameter[parameterIndex].getDeclaredType().toString();
        PhpClass expectedClass = PhpElementsUtil.getClassInterface(holder.getProject(), className);
        if(expectedClass == null) {
            return;
        }

        PhpClass serviceParameterClass = ServiceUtil.getResolvedClassDefinition(holder.getProject(), serviceName, lazyServiceCollector);
        if(serviceParameterClass != null && !PhpElementsUtil.isInstanceOf(serviceParameterClass, expectedClass)) {
            holder.registerProblem(
                target,
                "Expect instance of: " + expectedClass.getPresentableFQN(),
                ProblemHighlightType.GENERIC_ERROR_OR_WARNING,
                new XmlServiceSuggestIntentionAction(expectedClass.getFQN(), target)
            );
        }
    }
}

File: src/main/kotlin/fr/adrienbrault/idea/symfony2plugin/dic/inspection/XmlServiceInstanceInspection.kt
package fr.adrienbrault.idea.symfony2plugin.dic.inspection

import com.intellij.codeInspection.LocalInspectionTool
import com.intellij.codeInspection.ProblemHighlightType
import com.intellij.codeInspection.ProblemsHolder
import com.intellij.patterns.ElementPattern
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiElementVisitor
import com.intellij.psi.util.PsiTreeUtil
import com.intellij.psi.xml.XmlAttributeValue
import com.intellij.psi.xml.XmlTag
import com.jetbrains.php.lang.psi.elements.Method
import fr.adrienbrault.idea.symfony2plugin.Symfony2ProjectComponent
import fr.adrienbrault.idea.symfony2plugin.config.xml.XmlHelper
import fr.adrienbrault.idea.symfony2plugin.dic.container.suggestion.XmlServiceSuggestIntentionAction
import fr.adrienbrault.idea.symfony2plugin.stubs.ContainerCollectionResolver
import fr.adrienbrault.idea.symfony2plugin.util.PhpElementsUtil
import fr.adrienbrault.idea.symfony2plugin.util.dict.ServiceUtil
import org.apache.commons.lang3.StringUtils

private fun annotateServiceInstance(psiElement: PsiElement, holder: ProblemsHolder, lazyServiceCollector: ContainerCollectionResolver.LazyServiceCollector) {
    // search for parent service definition
    val currentXmlTag = PsiTreeUtil.getParentOfType(psiElement, XmlTag::class.java) ?: return
    val parentXmlTag = PsiTreeUtil.getParentOfType(currentXmlTag, XmlTag::class.java)
    if (parentXmlTag == null) {
        return
    }

    val name = parentXmlTag.name
    if (name == "service") {
        // service/argument[id]
        val serviceDefName = XmlHelper.getClassFromServiceDefinition(parentXmlTag)
        if (serviceDefName != null) {
            val phpClass = ServiceUtil.getResolvedClassDefinition(holder.project, serviceDefName, lazyServiceCollector)

            // check type hint on constructor
            if (phpClass != null) {
                val constructor = phpClass.constructor
                if (constructor != null) {
                    val serviceName = (psiElement as XmlAttributeValue).value
                    if (StringUtils.isNotBlank(serviceName)) {
                        attachMethodInstances(psiElement, serviceName, constructor, XmlHelper.getArgumentIndex(currentXmlTag, constructor), holder, lazyServiceCollector)
                    }
                }
            }
        }
    } else if (name == "call") {

        // service/call/argument[id]

        val methodAttribute = parentXmlTag.getAttribute("method")
        if (methodAttribute != null) {
            val methodName = methodAttribute.value
            val serviceTag = parentXmlTag.parentTag

            // get service class
            if (serviceTag != null && "service" == serviceTag.name) {
                val serviceDefName = XmlHelper.getClassFromServiceDefinition(serviceTag)
                if (serviceDefName != null) {
                    val phpClass = ServiceUtil.getResolvedClassDefinition(holder.project, serviceDefName, lazyServiceCollector)

                    // finally check method type hint
                    if (phpClass != null) {
                        val method = phpClass.findMethodByName(methodName)
                        if (method != null) {
                            val serviceName = (psiElement as XmlAttributeValue).value
                            if (StringUtils.isNotBlank(serviceName)) {
                                attachMethodInstances(psiElement, serviceName, method, XmlHelper.getArgumentIndex(currentXmlTag, method), holder, lazyServiceCollector)
                            }
                        }
                    }
                }
            }
        }
    }
}

private fun attachMethodInstances(target: PsiElement, serviceName: String, method: Method, parameterIndex: Int, holder: ProblemsHolder, lazyServiceCollector: ContainerCollectionResolver.LazyServiceCollector) {
    val constructorParameter = method.parameters
    if (parameterIndex >= constructorParameter.size) {
        return
    }

    val className = constructorParameter[parameterIndex].declaredType.toString()
    val expectedClass = PhpElementsUtil.getClassInterface(holder.project, className)
    if (expectedClass == null) {
        return
    }

    val serviceParameterClass = ServiceUtil.getResolvedClassDefinition(holder.project, serviceName, lazyServiceCollector)
    if (serviceParameterClass != null && !PhpElementsUtil.isInstanceOf(serviceParameterClass, expectedClass)) {
        holder.registerProblem(
            target,
            "Expect instance of: " + expectedClass.presentableFQN,
            ProblemHighlightType.GENERIC_ERROR_OR_WARNING,
            XmlServiceSuggestIntentionAction(expectedClass.fqn, target)
        )
    }
}

/**
 * @author Daniel Espendiller <daniel@espendiller.net>
 */
open class XmlServiceInstanceInspection : LocalInspectionTool() {
    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean): PsiElementVisitor {
        if (!Symfony2ProjectComponent.isEnabled(holder.project)) {
            return super.buildVisitor(holder, isOnTheFly)
        }

        return MyPsiElementVisitor(holder)
    }

    private class MyPsiElementVisitor(private val holder: ProblemsHolder) : PsiElementVisitor() {
        private var lazyServiceCollector: ContainerCollectionResolver.LazyServiceCollector? = null
        private var argumentServiceIdPattern: ElementPattern<*>? = null

        override fun visitElement(psiElement: PsiElement) {
            if (getArgumentServiceIdPattern().accepts(psiElement)) {
                val collector = lazyServiceCollector ?: ContainerCollectionResolver.LazyServiceCollector(holder.project)
                lazyServiceCollector = collector

                annotateServiceInstance(psiElement, holder, collector)
            }

            super.visitElement(psiElement)
        }

        private fun getArgumentServiceIdPattern(): ElementPattern<*> {
            val pattern = argumentServiceIdPattern ?: XmlHelper.getArgumentServiceIdPattern()
            argumentServiceIdPattern = pattern
            return pattern
        }
    }

}
